File: src/baekjoon/step/graph/PROB1260.java
package baekjoon.step.graph;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class PROB1260 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();
    static int N, M, R;
    static ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
    static Queue<Integer> queue = new LinkedList<>();
    static boolean[] dfsVisited;
    static boolean[] bfsVisited;
    static int dfsCnt = 1;
    static int bfsCnt = 1;

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        dfsVisited = new boolean[N + 1];
        bfsVisited = new boolean[N + 1];

        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());

            graph.get(u).add(v);
            graph.get(v).add(u);
        }

        for (int i = 1; i <= N; i++) {
            Collections.sort(graph.get(i));
        }

        dfs(R);
        sb.append("\n");
        bfs(R);
        System.out.println(sb);
    }

    public static void bfs(int startNode) {
        queue.add(startNode);
        bfsVisited[startNode] = true;

        while (!queue.isEmpty()) {
            int currentNode = queue.poll();
            sb.append(currentNode).append(" ");

            for (int nextNode : graph.get(currentNode)) {
                if (!bfsVisited[nextNode]) {
                    bfsVisited[nextNode] = true;
                    queue.add(nextNode);
                }
            }
        }
    }

    public static void dfs(int startNode) {
        dfsVisited[startNode] = true;
        sb.append(startNode).append(" ");

        for (Integer nextNode : graph.get(startNode)) {
            if (!dfsVisited[nextNode]) {
                dfs(nextNode);
            }
        }
    }
}

File: src/baekjoon/step/graph/PROB1260.py
# 그래프를 DFS 로 탐색한 결과와 BFS 로 탐색한 결과를 출력하는 프로그램을 작성하시오.
# 정점 번호가 작은 곳부터 먼저 방문한다.


import sys
from collections import deque
sys.setrecursionlimit(10**6)


# 정점의 개수, 간선의 개수, 탐색을 시작할 정점의 번호 입력받기
N, M, V = map(int, input().split())

# 그래프 만들기
graph = [[] for i in range(N + 1)]

for i in range(M) :
    v, u = map(int, sys.stdin.readline().split())
    graph[u].append(v)
    graph[v].append(u)

# 그래프 정렬
for i in graph :
    i.sort()

################################################################

# 방문 리스트
visited_dfs = [False] * (N + 1)

# DFS 함수
def dfs(v) :
    visited_dfs[v] = True
    print(v, end=' ')

    for i in graph[v] :
        if not visited_dfs[i] :
            dfs(i)

dfs(V)
print()

################################################################


# 방문 리스트
visited_bfs = [0] * (N + 1)

# BFS 함수
def bfs(v) :
    # 큐 구현
    queue = deque([v])
    visited_bfs[v] = True

    while queue :
        v = queue.popleft()
        print(v, end=' ')

        for i in graph[v] :
            if not visited_bfs[i] :
                visited_bfs[i] = True
                queue.append(i)

bfs(V)
